Crash while building/rebuiding View

Greetings,

On my Galaxy S2 I am receiving the following crash consistently:
```
03-16 11:53:52.488  20863-20975/com.app.appmobile.alpha W/CursorWindow﹕ Window is full: requested allocation 39055 bytes, free space 4069 bytes, window size 2097152 bytes
03-16 11:53:54.450  20863-20975/com.app.appmobile.alpha W/CursorWindow﹕ Window is full: requested allocation 39055 bytes, free space 621 bytes, window size 2097152 bytes
03-16 11:53:54.781  20863-20975/com.app.appmobile.alpha W/CursorWindow﹕ Window is full: requested allocation 39055 bytes, free space 10177 bytes, window size 2097152 bytes
03-16 11:53:54.791  20863-20975/com.app.appmobile.alpha E/CursorWindow﹕ Failed to read row 486, column 0 from a CursorWindow which has 486 rows, 7 columns.
03-16 11:53:54.791  20863-20975/com.app.appmobile.alpha W/View﹕ Failed to rebuild view users.  Result code: 500
03-16 11:53:54.791  20863-20975/com.app.appmobile.alpha I/CBLite﹕ CBLManagerWorkExecutor CANCEL transaction (level 0)
03-16 11:53:54.821  20863-20975/com.app.appmobile.alpha E/Sync﹕ Exception calling changeListener.changed
    java.lang.IllegalStateException: Couldn't read row 486, col 0 from CursorWindow.  Make sure the Cursor is initialized correctly before accessing data from it.
            at android.database.CursorWindow.nativeGetLong(Native Method)
            at android.database.CursorWindow.getLong(CursorWindow.java:511)
            at android.database.AbstractWindowedCursor.getLong(AbstractWindowedCursor.java:75)
            at com.couchbase.lite.android.AndroidSQLiteStorageEngine$SQLiteCursorWrapper.getLong(AndroidSQLiteStorageEngine.java:198)
            at com.couchbase.lite.View.updateIndex(View.java:543)
            at com.couchbase.lite.Database.queryViewNamed(Database.java:2204)
            at com.couchbase.lite.Query.run(Query.java:395)
            at com.app.applib.model.user.User.getUser(User.java:55)
            at com.app.appmobile.AppEngine.SyncComplete(AppEngine.java:162)
            at com.app.applib.network.Sync.evaluateSessionStatus(Sync.java:304)
            at com.app.applib.network.Sync.changed(Sync.java:243)
            at com.couchbase.lite.replicator.Replication.changed(Replication.java:357)
            at com.couchbase.lite.replicator.ReplicationInternal$10.run(ReplicationInternal.java:1028)
            at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:442)
            at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305)
            at java.util.concurrent.FutureTask.run(FutureTask.java:137)
            at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:150)
            at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:264)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
            at java.lang.Thread.run(Thread.java:856)
```

I've tried a simple background looper waiting until view.isStale() is false before utilizing the view query, but view.isStale() always returns true and the lastSequenceIndexed() is always zero.

The view is as follows:
```
public void createUsersView() {
        if (database == null) { return; }
        View view = database.getView("users");
        Mapper mapper = new Mapper() {
            @Override
            public void map(Map<String, Object> doc, Emitter emitter) {
                if (doc.get("type") != null && doc.get("type").equals("user")) {
                    emitter.emit(doc.get("name"), doc.get("locationRoot"));
                }
            }
        };
        view.setMap(mapper, "1.0");
        usersQuery = view.createQuery();
    }
```

On an emulator I am *not* receiving this error or crashing, so I'm guessing it has something to do with this device's slower processor or something hardware related.

Thanks,
Justin

Log Message

Android Standard Emulator with x86, 1G memory, API 16 configuration
```
3-23 16:40:30.541    1549-1549/self.hideki.issue497 E/CursorWindow﹕ Failed to read row 151, column 0 from a CursorWindow which has 151 rows, 7 columns.
03-23 16:40:30.541    1549-1549/self.hideki.issue497 W/View﹕ Failed to rebuild view users.  Result code: 500
03-23 16:40:30.541    1549-1549/self.hideki.issue497 D/AndroidRuntime﹕ Shutting down VM
03-23 16:40:30.541    1549-1549/self.hideki.issue497 W/dalvikvm﹕ threadid=1: thread exiting with uncaught exception (group=0xb2e28288)
03-23 16:40:30.541    1549-1549/self.hideki.issue497 E/AndroidRuntime﹕ FATAL EXCEPTION: main
    java.lang.IllegalStateException: Could not execute method of the activity
            at android.view.View$1.onClick(View.java:3591)
            at android.view.View.performClick(View.java:4084)
            at android.view.View$PerformClick.run(View.java:16966)
            at android.os.Handler.handleCallback(Handler.java:615)
            at android.os.Handler.dispatchMessage(Handler.java:92)
            at android.os.Looper.loop(Looper.java:137)
            at android.app.ActivityThread.main(ActivityThread.java:4745)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
            at dalvik.system.NativeStart.main(Native Method)
     Caused by: java.lang.reflect.InvocationTargetException
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at android.view.View$1.onClick(View.java:3586)
            at android.view.View.performClick(View.java:4084)
            at android.view.View$PerformClick.run(View.java:16966)
            at android.os.Handler.handleCallback(Handler.java:615)
            at android.os.Handler.dispatchMessage(Handler.java:92)
            at android.os.Looper.loop(Looper.java:137)
            at android.app.ActivityThread.main(ActivityThread.java:4745)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
            at dalvik.system.NativeStart.main(Native Method)
     Caused by: java.lang.IllegalStateException: Couldn't read row 151, col 0 from CursorWindow.  Make sure the Cursor is initialized correctly before accessing data from it.
            at android.database.CursorWindow.nativeGetLong(Native Method)
            at android.database.CursorWindow.getLong(CursorWindow.java:507)
            at android.database.AbstractWindowedCursor.getLong(AbstractWindowedCursor.java:75)
            at com.couchbase.lite.android.AndroidSQLiteStorageEngine$SQLiteCursorWrapper.getLong(AndroidSQLiteStorageEngine.java:215)
            at com.couchbase.lite.View.updateIndex(View.java:541)
            at com.couchbase.lite.Database.queryViewNamed(Database.java:2228)
            at com.couchbase.lite.Query.run(Query.java:395)
            at self.hideki.issue497.Application.iterateQuery(Application.java:95)
            at self.hideki.issue497.MainActivity.iterateQuery(MainActivity.java:62)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at android.view.View$1.onClick(View.java:3586)
            at android.view.View.performClick(View.java:4084)
            at android.view.View$PerformClick.run(View.java:16966)
            at android.os.Handler.handleCallback(Handler.java:615)
            at android.os.Handler.dispatchMessage(Handler.java:92)
            at android.os.Looper.loop(Looper.java:137)
            at android.app.ActivityThread.main(ActivityThread.java:4745)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
            at dalvik.system.NativeStart.main(Native Method)

```

Another Log from Android Xoom 1GB memory device
```
03-23 16:56:04.810  15411-15411/self.hideki.issue497 E/Issue447﹕ iterateQuery() is clicked
03-23 16:56:05.060  15411-15411/self.hideki.issue497 W/CursorWindow﹕ Window is full: requested allocation 40496 bytes, free space 21077 bytes, window size 2097152 bytes
03-23 16:56:05.300  15411-15415/self.hideki.issue497 D/dalvikvm﹕ GC_CONCURRENT freed 1044K, 73% free 8221K/29383K, paused 12ms+9ms, total 67ms
03-23 16:56:05.300  15411-15411/self.hideki.issue497 D/dalvikvm﹕ WAIT_FOR_CONCURRENT_GC blocked 26ms
03-23 16:56:05.460  15411-15415/self.hideki.issue497 D/dalvikvm﹕ GC_CONCURRENT freed 817K, 73% free 8153K/29383K, paused 3ms+6ms, total 48ms
03-23 16:56:05.460  15411-15411/self.hideki.issue497 D/dalvikvm﹕ WAIT_FOR_CONCURRENT_GC blocked 23ms
03-23 16:56:05.580  15411-15415/self.hideki.issue497 D/dalvikvm﹕ GC_CONCURRENT freed 626K, 72% free 8253K/29383K, paused 2ms+6ms, total 44ms
03-23 16:56:05.580  15411-15411/self.hideki.issue497 D/dalvikvm﹕ WAIT_FOR_CONCURRENT_GC blocked 33ms
03-23 16:56:05.790  15411-15411/self.hideki.issue497 W/CursorWindow﹕ Window is full: requested allocation 40336 bytes, free space 25417 bytes, window size 2097152 bytes
03-23 16:56:05.960  15411-15411/self.hideki.issue497 W/CursorWindow﹕ Window is full: requested allocation 40336 bytes, free space 27721 bytes, window size 2097152 bytes
03-23 16:56:05.960  15411-15411/self.hideki.issue497 E/CursorWindow﹕ Failed to read row 151, column 0 from a CursorWindow which has 151 rows, 7 columns.
03-23 16:56:05.960  15411-15411/self.hideki.issue497 W/View﹕ Failed to rebuild view users.  Result code: 500
03-23 16:56:05.970  15411-15411/self.hideki.issue497 D/AndroidRuntime﹕ Shutting down VM
03-23 16:56:05.970  15411-15411/self.hideki.issue497 W/dalvikvm﹕ threadid=1: thread exiting with uncaught exception (group=0x41de3300)
03-23 16:56:05.990  15411-15411/self.hideki.issue497 E/AndroidRuntime﹕ FATAL EXCEPTION: main
    java.lang.IllegalStateException: Could not execute method of the activity
            at android.view.View$1.onClick(View.java:3591)
            at android.view.View.performClick(View.java:4084)
            at android.view.View$PerformClick.run(View.java:16966)
            at android.os.Handler.handleCallback(Handler.java:615)
            at android.os.Handler.dispatchMessage(Handler.java:92)
            at android.os.Looper.loop(Looper.java:137)
            at android.app.ActivityThread.main(ActivityThread.java:4745)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
            at dalvik.system.NativeStart.main(Native Method)
     Caused by: java.lang.reflect.InvocationTargetException
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at android.view.View$1.onClick(View.java:3586)
            at android.view.View.performClick(View.java:4084)
            at android.view.View$PerformClick.run(View.java:16966)
            at android.os.Handler.handleCallback(Handler.java:615)
            at android.os.Handler.dispatchMessage(Handler.java:92)
            at android.os.Looper.loop(Looper.java:137)
            at android.app.ActivityThread.main(ActivityThread.java:4745)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
            at dalvik.system.NativeStart.main(Native Method)
     Caused by: java.lang.IllegalStateException: Couldn't read row 151, col 0 from CursorWindow.  Make sure the Cursor is initialized correctly before accessing data from it.
            at android.database.CursorWindow.nativeGetLong(Native Method)
            at android.database.CursorWindow.getLong(CursorWindow.java:507)
            at android.database.AbstractWindowedCursor.getLong(AbstractWindowedCursor.java:75)
            at com.couchbase.lite.android.AndroidSQLiteStorageEngine$SQLiteCursorWrapper.getLong(AndroidSQLiteStorageEngine.java:215)
            at com.couchbase.lite.View.updateIndex(View.java:541)
            at com.couchbase.lite.Database.queryViewNamed(Database.java:2228)
            at com.couchbase.lite.Query.run(Query.java:395)
            at self.hideki.issue497.Application.iterateQuery(Application.java:95)
            at self.hideki.issue497.MainActivity.iterateQuery(MainActivity.java:62)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at android.view.View$1.onClick(View.java:3586)
            at android.view.View.performClick(View.java:4084)
            at android.view.View$PerformClick.run(View.java:16966)
            at android.os.Handler.handleCallback(Handler.java:615)
            at android.os.Handler.dispatchMessage(Handler.java:92)
            at android.os.Looper.loop(Looper.java:137)
            at android.app.ActivityThread.main(ActivityThread.java:4745)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:511)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:786)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:553)
            at dalvik.system.NativeStart.main(Native Method)
```

--- a/src/main/java/com/couchbase/lite/View.java
+++ b/src/main/java/com/couchbase/lite/View.java
@@ -523,6 +523,14 @@ public final class View {
 
             boolean keepGoing = cursor.moveToNext();
             while (keepGoing) {
+
+                // NOTE: skip row if 1st column is null
+                // https://github.com/couchbase/couchbase-lite-java-core/issues/497
+                if (cursor.isNull(0)) {
+                    keepGoing = cursor.moveToNext();
+                    continue;
+                }
+
                 long docID = cursor.getLong(0);
 
                 // Reconstitute the document as a dictionary:
@@ -538,8 +546,11 @@ public final class View {
                 boolean noAttachments = cursor.getInt(5) > 0;
                 boolean deleted = cursor.getInt(6) > 0;
 
-                while ((keepGoing = cursor.moveToNext()) &&  cursor.getLong(0) == docID) {
+                while ((keepGoing = cursor.moveToNext()) && (cursor.isNull(0) || cursor.getLong(0) == docID)) {
                     // Skip rows with the same doc_id -- these are losing conflicts.
+
+                    // NOTE: Or Skip rows if 1st column is null
+                    // https://github.com/couchbase/couchbase-lite-java-core/issues/497
                 }
 
                 if (minLastSequence > 0) {

--- a/src/main/java/com/couchbase/lite/storage/Cursor.java
+++ b/src/main/java/com/couchbase/lite/storage/Cursor.java
@@ -24,4 +24,5 @@ public interface Cursor {
     long getLong(int columnIndex);
     byte[] getBlob(int columnIndex);
     void close();
+    boolean isNull(int columnIndex);
 }

